lib/mix/lib/mix/tasks/archive.install.ex

```diff
@@ -104,6 +105,18 @@ defmodule Mix.Tasks.Archive.Install do
 
   ### Private helpers
 
+  defp destination(ez_dst, contents) do
+    with {:ok, [_comment, zip_first_file | _]} <- :zip.list_dir(contents),
+         {:zip_file, zip_first_path, _, _, _, _} <- zip_first_file,
+         [zip_root_dir | _] <- Path.split(zip_first_path) do
+
+      {:ok, Path.join(Path.dirname(ez_dst), zip_root_dir)}
+    else
+      _ ->
+        {:error, "Invalid archive file"}
```

Comment:
Messages from other places used verbatim though, let's do the opposite then and move `Installation failed:` here?